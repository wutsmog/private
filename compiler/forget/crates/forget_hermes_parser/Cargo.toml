[package]
name = "forget_hermes_parser"
version = "0.1.0"
authors.workspace = true
description.workspace = true
edition.workspace = true
homepage.workspace = true
keywords.workspace = true
license.workspace = true
repository.workspace = true

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
insta = { workspace = true }
forget_diagnostics = { workspace = true }
forget_estree = { workspace = true }
hermes = { workspace = true } 
juno_support = { workspace = true }
serde_json = { workspace = true }

[build-dependencies]
forget_estree_codegen = { workspace = true }