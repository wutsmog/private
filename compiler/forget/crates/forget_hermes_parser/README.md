# forget_hermes_parser

Wrapper around the Hermes parser that exposes parse results as a `forget_estree` AST.