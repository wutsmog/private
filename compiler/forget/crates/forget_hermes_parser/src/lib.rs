mod generated;
mod generated_extension;

use forget_diagnostics::Diagnostic;
use forget_estree::Program;
use generated_extension::{Context, FromHermes};
use hermes::parser::{HermesParser, ParserDialect, ParserFlags};
use hermes::utf::utf8_with_surrogates_to_string;
use juno_support::NullTerminatedBuf;

pub fn parse(source: &str, _file: &str) -> Result<Program, Vec<Diagnostic>> {
    let buf = NullTerminatedBuf::from_str_check(source);
    let result = HermesParser::parse(
        ParserFlags {
            dialect: ParserDialect::TypeScript,
            enable_jsx: true,
            store_doc_block: true,
            strict_mode: true,
        },
        &buf,
    );
    let mut cx = Context::new();
    if result.has_errors() {
        let error_messages = result.messages();
        return Err(error_messages
            .into_iter()
            .map(|diag| {
                let message = utf8_with_surrogates_to_string(diag.message.as_slice()).unwrap();
                Diagnostic::invalid_syntax(message, None)
            })
            .collect());
    }

    Ok(FromHermes::convert(&mut cx, result.root().unwrap()))
}
