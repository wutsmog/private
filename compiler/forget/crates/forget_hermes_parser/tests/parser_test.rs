use std::env;

use forget_hermes_parser::parse;
use insta::{assert_snapshot, glob};
use serde_json;

#[test]
fn fixtures() {
    glob!("fixtures/**.js", |path| {
        println!("fixture {}", path.to_str().unwrap());
        let input = std::fs::read_to_string(path).unwrap();
        let ast = parse(&input, path.to_str().unwrap()).unwrap();
        let output = serde_json::to_string_pretty(&ast).unwrap();
        let output = output.trim();
        assert_snapshot!(format!("Input:\n{input}\n\nOutput:\n{output}"));
    });
}
