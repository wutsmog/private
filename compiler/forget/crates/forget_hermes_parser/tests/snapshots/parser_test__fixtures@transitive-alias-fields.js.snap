---
source: crates/forget_hermes_parser/tests/parser_test.rs
expression: "format!(\"Input:\\n{input}\\n\\nOutput:\\n{output}\")"
input_file: crates/forget_hermes_parser/tests/fixtures/transitive-alias-fields.js
---
Input:
function component() {
  let x = {};
  let p = {};
  let q = {};
  let y = {};

  x.y = y;
  p.y = x.y;
  q.y = p.y;

  mutate(q);
}


Output:
{
  "type": "Program",
  "body": [
    {
      "type": "FunctionDeclaration",
      "id": {
        "type": "Identifier",
        "name": "component",
        "typeAnnotation": null,
        "loc": null,
        "range": {
          "start": 9,
          "end": 18
        }
      },
      "params": [],
      "body": {
        "type": "BlockStatement",
        "body": [
          {
            "type": "VariableDeclaration",
            "kind": "let",
            "declarations": [
              {
                "type": "VariableDeclarator",
                "id": {
                  "type": "Identifier",
                  "name": "x",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": {
                    "start": 29,
                    "end": 30
                  }
                },
                "init": {
                  "type": "ObjectExpression",
                  "properties": [],
                  "loc": null,
                  "range": {
                    "start": 33,
                    "end": 35
                  }
                },
                "loc": null,
                "range": {
                  "start": 29,
                  "end": 35
                }
              }
            ],
            "loc": null,
            "range": {
              "start": 25,
              "end": 36
            }
          },
          {
            "type": "VariableDeclaration",
            "kind": "let",
            "declarations": [
              {
                "type": "VariableDeclarator",
                "id": {
                  "type": "Identifier",
                  "name": "p",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": {
                    "start": 43,
                    "end": 44
                  }
                },
                "init": {
                  "type": "ObjectExpression",
                  "properties": [],
                  "loc": null,
                  "range": {
                    "start": 47,
                    "end": 49
                  }
                },
                "loc": null,
                "range": {
                  "start": 43,
                  "end": 49
                }
              }
            ],
            "loc": null,
            "range": {
              "start": 39,
              "end": 50
            }
          },
          {
            "type": "VariableDeclaration",
            "kind": "let",
            "declarations": [
              {
                "type": "VariableDeclarator",
                "id": {
                  "type": "Identifier",
                  "name": "q",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": {
                    "start": 57,
                    "end": 58
                  }
                },
                "init": {
                  "type": "ObjectExpression",
                  "properties": [],
                  "loc": null,
                  "range": {
                    "start": 61,
                    "end": 63
                  }
                },
                "loc": null,
                "range": {
                  "start": 57,
                  "end": 63
                }
              }
            ],
            "loc": null,
            "range": {
              "start": 53,
              "end": 64
            }
          },
          {
            "type": "VariableDeclaration",
            "kind": "let",
            "declarations": [
              {
                "type": "VariableDeclarator",
                "id": {
                  "type": "Identifier",
                  "name": "y",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": {
                    "start": 71,
                    "end": 72
                  }
                },
                "init": {
                  "type": "ObjectExpression",
                  "properties": [],
                  "loc": null,
                  "range": {
                    "start": 75,
                    "end": 77
                  }
                },
                "loc": null,
                "range": {
                  "start": 71,
                  "end": 77
                }
              }
            ],
            "loc": null,
            "range": {
              "start": 67,
              "end": 78
            }
          },
          {
            "type": "ExpressionStatement",
            "expression": {
              "type": "AssignmentExpression",
              "operator": "=",
              "left": {
                "type": "MemberExpression",
                "object": {
                  "type": "Identifier",
                  "name": "x",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": {
                    "start": 82,
                    "end": 83
                  }
                },
                "property": {
                  "type": "Identifier",
                  "name": "y",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": {
                    "start": 84,
                    "end": 85
                  }
                },
                "computed": false,
                "loc": null,
                "range": {
                  "start": 82,
                  "end": 85
                }
              },
              "right": {
                "type": "Identifier",
                "name": "y",
                "typeAnnotation": null,
                "loc": null,
                "range": {
                  "start": 88,
                  "end": 89
                }
              },
              "loc": null,
              "range": {
                "start": 82,
                "end": 89
              }
            },
            "directive": null,
            "loc": null,
            "range": {
              "start": 82,
              "end": 90
            }
          },
          {
            "type": "ExpressionStatement",
            "expression": {
              "type": "AssignmentExpression",
              "operator": "=",
              "left": {
                "type": "MemberExpression",
                "object": {
                  "type": "Identifier",
                  "name": "p",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": {
                    "start": 93,
                    "end": 94
                  }
                },
                "property": {
                  "type": "Identifier",
                  "name": "y",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": {
                    "start": 95,
                    "end": 96
                  }
                },
                "computed": false,
                "loc": null,
                "range": {
                  "start": 93,
                  "end": 96
                }
              },
              "right": {
                "type": "MemberExpression",
                "object": {
                  "type": "Identifier",
                  "name": "x",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": {
                    "start": 99,
                    "end": 100
                  }
                },
                "property": {
                  "type": "Identifier",
                  "name": "y",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": {
                    "start": 101,
                    "end": 102
                  }
                },
                "computed": false,
                "loc": null,
                "range": {
                  "start": 99,
                  "end": 102
                }
              },
              "loc": null,
              "range": {
                "start": 93,
                "end": 102
              }
            },
            "directive": null,
            "loc": null,
            "range": {
              "start": 93,
              "end": 103
            }
          },
          {
            "type": "ExpressionStatement",
            "expression": {
              "type": "AssignmentExpression",
              "operator": "=",
              "left": {
                "type": "MemberExpression",
                "object": {
                  "type": "Identifier",
                  "name": "q",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": {
                    "start": 106,
                    "end": 107
                  }
                },
                "property": {
                  "type": "Identifier",
                  "name": "y",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": {
                    "start": 108,
                    "end": 109
                  }
                },
                "computed": false,
                "loc": null,
                "range": {
                  "start": 106,
                  "end": 109
                }
              },
              "right": {
                "type": "MemberExpression",
                "object": {
                  "type": "Identifier",
                  "name": "p",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": {
                    "start": 112,
                    "end": 113
                  }
                },
                "property": {
                  "type": "Identifier",
                  "name": "y",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": {
                    "start": 114,
                    "end": 115
                  }
                },
                "computed": false,
                "loc": null,
                "range": {
                  "start": 112,
                  "end": 115
                }
              },
              "loc": null,
              "range": {
                "start": 106,
                "end": 115
              }
            },
            "directive": null,
            "loc": null,
            "range": {
              "start": 106,
              "end": 116
            }
          },
          {
            "type": "ExpressionStatement",
            "expression": {
              "type": "CallExpression",
              "callee": {
                "type": "Identifier",
                "name": "mutate",
                "typeAnnotation": null,
                "loc": null,
                "range": {
                  "start": 120,
                  "end": 126
                }
              },
              "arguments": [
                {
                  "type": "Identifier",
                  "name": "q",
                  "typeAnnotation": null,
                  "loc": null,
                  "range": {
                    "start": 127,
                    "end": 128
                  }
                }
              ],
              "loc": null,
              "range": {
                "start": 120,
                "end": 129
              }
            },
            "directive": null,
            "loc": null,
            "range": {
              "start": 120,
              "end": 130
            }
          }
        ],
        "loc": null,
        "range": {
          "start": 21,
          "end": 132
        }
      },
      "generator": false,
      "async": false,
      "loc": null,
      "range": {
        "start": 0,
        "end": 132
      },
      "loc": null,
      "range": {
        "start": 0,
        "end": 132
      }
    }
  ],
  "sourceType": "script",
  "loc": null,
  "range": {
    "start": 0,
    "end": 132
  }
}
